File: src/systems/vfx/vfxProgram.ts
/** Authoring -> runtime boundary. All historical defaults are resolved here.
 * No asset mutation, random draws, simulation state, or UI dependencies.
 */
import type { VfxEmitterDef, VfxSimulationDef } from '../../data/types';
import contract from '../../data/vfxSimulationContract.json';

export type VfxEmitterProgram = Readonly<VfxSimulationDef>;

/** The effective old behavior, including previously inactive authored fields.
 * Materializing this plan is a lossless authoring operation, not an upgrade of physics.
 */
export function resolveEmitterProgram(def: VfxEmitterDef): VfxEmitterProgram {
  if (def.simulation) return { initialVelocity: 'configured', ...structuredClone(def.simulation) };
  const solver = def.behavior ? 'flock' : def.plate ? 'plate' : 'particle';
  return {
    solver,
    spawnPlacement: solver === 'plate' && def.spawn.shape?.kind === 'area' ? 'surface' : 'shape',
    initialVelocity: solver === 'plate' && def.spawn.shape?.kind === 'area' ? 'rest' : 'configured',
    influences: {
      sceneWind: solver !== 'flock', wind: solver === 'particle', airflow: false,
      stimulus: solver !== 'plate',
    },
    recycle: { mode: solver === 'plate' && def.plate?.replenish !== false ? 'airborne' : 'none' },
  };
}

/** New emitters opt into the public input contract; existing emitters use resolve above. */
export function newEmitterProgram(solver: VfxSimulationDef['solver']): VfxSimulationDef {
  return structuredClone(contract.newPrograms[solver]) as VfxSimulationDef;
}

/** Model switches reset known execution options but preserve unknown authored data. */
export function switchEmitterProgram(def: VfxEmitterDef, solver: VfxSimulationDef['solver']): VfxSimulationDef {
  const previous = resolveEmitterProgram(def), next = newEmitterProgram(solver);
  return { ...previous, ...next, influences: { ...previous.influences, ...next.influences }, recycle: { ...previous.recycle, ...next.recycle } };
}

/** Shared by runtime construction and the workbench. Invalid combinations cannot silently fall through. */
export function emitterProgramErrors(def: VfxEmitterDef): string[] {
  if (def.simulation === undefined) return []; // Existing documents retain their exact interpretation.
  if (!def.simulation || typeof def.simulation !== 'object' || Array.isArray(def.simulation)) return ['simulation 必须为对象'];
  const p = def.simulation, errors: string[] = [];
  if (!contract.solvers.includes(p.solver)) errors.push('simulation.solver 无效');
  if (!contract.spawnPlacements.includes(p.spawnPlacement)) errors.push('simulation.spawnPlacement 无效');
  if (p.surfaceRadius !== undefined && (!Number.isFinite(p.surfaceRadius) || p.surfaceRadius <= 0)) errors.push('simulation.surfaceRadius 必须为正数');
  if (p.initialVelocity !== undefined && !contract.initialVelocities.includes(p.initialVelocity)) errors.push('simulation.initialVelocity 无效');
  for (const k of contract.influenceKeys as (keyof VfxSimulationDef['influences'])[]) {
    if (typeof p.influences?.[k] !== 'boolean') errors.push(`simulation.influences.${k} 必须为布尔`);
  }
  for (const k of contract.optionalInfluenceKeys as (keyof VfxSimulationDef['influences'])[]) {
    if (p.influences?.[k] !== undefined && typeof p.influences[k] !== 'boolean') errors.push(`simulation.influences.${k} 必须为布尔`);
  }
  if (!contract.recycleModes.includes(p.recycle?.mode)) errors.push('simulation.recycle.mode 无效');
  for (const k of ['height', 'upwind'] as const) {
    const v = p.recycle?.[k];
    if (v !== undefined && (!Array.isArray(v) || v.length !== 2 || !v.every(n => Number.isFinite(n) && n >= 0) || v[0] > v[1])) {
      errors.push(`simulation.recycle.${k} 必须为非负递增区间`);
    }
  }
  if (p.solver === 'plate' && (!def.plate || typeof def.plate !== 'object' || Array.isArray(def.plate))) errors.push('薄片求解器缺少 plate 参数');
  if (p.solver === 'flock' && (!def.behavior || typeof def.behavior !== 'object' || Array.isArray(def.behavior))) errors.push('群体求解器缺少 behavior 参数');
  if (p.solver === 'flock' && (def.subOnly || p.spawnPlacement !== 'shape' || p.recycle?.mode !== 'none')) {
    errors.push('群体由巢管理出生与返回，不能作为子发射器或使用表面铺撒 / 补回');
  }
  if (p.solver === 'flock' && (p.influences?.sceneWind || p.influences?.wind || p.influences?.airflow)) {
    errors.push('群体运动模型不支持物理风输入；使用群体刺激响应');
  }
  if (p.solver === 'flock' && p.influences?.contact) errors.push('群体运动模型不支持接触冲量；使用群体刺激响应');
  return errors;
}

/** Effective capabilities for authoring. Inactive source fields stay in the document. */
export function emitterCapabilities(def: VfxEmitterDef) {
  const p = resolveEmitterProgram(def);
  return {
    ...p,
    initialVelocity: p.initialVelocity !== 'rest' && p.solver !== 'flock',
    genericMotion: p.solver === 'particle',
    turbulence: p.solver !== 'flock',
    collision: p.solver === 'particle',
    plate: p.solver === 'plate', flock: p.solver === 'flock',
    errors: emitterProgramErrors(def),
  };
}

File: tools/editor/shared/vfx_program.py
"""Explicit VFX execution configuration. Enum/default authority is the runtime JSON contract.

Validation mirrors vfxProgram.emitterProgramErrors; parity is verified with the same fixtures.
This module never modifies documents or writes files.
"""
from __future__ import annotations

import copy
import json
import math
from pathlib import Path

CONTRACT = json.loads((Path(__file__).resolve().parents[3] / "src/data/vfxSimulationContract.json").read_text(encoding="utf-8"))


def effective_solver(emitter: dict) -> str:
    program = emitter.get("simulation")
    if isinstance(program, dict):
        return program.get("solver", "")
    return "flock" if emitter.get("behavior") else "plate" if emitter.get("plate") else "particle"


def new_program(solver: str = "particle") -> dict:
    return copy.deepcopy(CONTRACT["newPrograms"][solver])


def program_errors(emitter: dict) -> list[str]:
    if "simulation" not in emitter:
        return []
    p = emitter["simulation"]
    if not isinstance(p, dict):
        return ["simulation 必须为对象"]
    errors = []
    solver = p.get("solver")
    if solver not in CONTRACT["solvers"]:
        errors.append("simulation.solver 无效")
    if p.get("spawnPlacement") not in CONTRACT["spawnPlacements"]:
        errors.append("simulation.spawnPlacement 无效")
    radius = p.get("surfaceRadius")
    if "surfaceRadius" in p and not (isinstance(radius, (int, float)) and not isinstance(radius, bool) and math.isfinite(radius) and radius > 0):
        errors.append("simulation.surfaceRadius 必须为正数")
    if "initialVelocity" in p and p["initialVelocity"] not in CONTRACT["initialVelocities"]:
        errors.append("simulation.initialVelocity 无效")
    inf = p.get("influences") if isinstance(p.get("influences"), dict) else {}
    for k in CONTRACT["influenceKeys"]:
        if not isinstance(inf.get(k), bool):
            errors.append(f"simulation.influences.{k} 必须为布尔")
    for k in CONTRACT["optionalInfluenceKeys"]:
        if k in inf and not isinstance(inf[k], bool):
            errors.append(f"simulation.influences.{k} 必须为布尔")
    recycle = p.get("recycle") if isinstance(p.get("recycle"), dict) else {}
    if recycle.get("mode") not in CONTRACT["recycleModes"]:
        errors.append("simulation.recycle.mode 无效")
    for k in ("height", "upwind"):
        if k not in recycle:
            continue
        v = recycle[k]
        if not (isinstance(v, list) and len(v) == 2
                and all(isinstance(n, (int, float)) and not isinstance(n, bool) and math.isfinite(n) and n >= 0 for n in v)
                and v[0] <= v[1]):
            errors.append(f"simulation.recycle.{k} 必须为非负递增区间")
    if solver == "plate" and not isinstance(emitter.get("plate"), dict):
        errors.append("薄片求解器缺少 plate 参数")
    if solver == "flock" and not isinstance(emitter.get("behavior"), dict):
        errors.append("群体求解器缺少 behavior 参数")
    if solver == "flock" and (emitter.get("subOnly") or p.get("spawnPlacement") != "shape" or recycle.get("mode") != "none"):
        errors.append("群体由巢管理出生与返回，不能作为子发射器或使用表面铺撒 / 补回")
    if solver == "flock" and any(inf.get(k) for k in ("sceneWind", "wind", "airflow")):
        errors.append("群体运动模型不支持物理风输入；使用群体刺激响应")
    if solver == "flock" and inf.get("contact"):
        errors.append("群体运动模型不支持接触冲量；使用群体刺激响应")
    return errors
